sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java

```diff
@@ -4739,4 +5025,424 @@ static UUID randomUuid(long msb, long lsb) {
+            nonNullRequestOptions);
+
+        if (orderedApplicableRegionsForSpeculation.size() < 2) {
+            // There is at most one applicable region - no hedging possible
+            return callback.apply(nonNullRequestOptions, endToEndPolicyConfig, null);
+        }
+
+        ThresholdBasedAvailabilityStrategy availabilityStrategy =
+            (ThresholdBasedAvailabilityStrategy)endToEndPolicyConfig.getAvailabilityStrategy();
+        List<Mono<NonTransientPointOperationResult>> monoList = new ArrayList<>();
+
+        final ScopedDiagnosticsFactory diagnosticsFactory = new ScopedDiagnosticsFactory(this);
+
+        orderedApplicableRegionsForSpeculation
+            .forEach(region -> {
+                RequestOptions clonedOptions = new RequestOptions(nonNullRequestOptions);
+
+                if (monoList.isEmpty()) {
+                    // no special error handling for transient errors to suppress them here
+                    // because any cross-regional retries are expected to be processed
+                    // by the ClientRetryPolicy for the initial request - so, any outcome of the
+                    // initial Mono should be treated as non-transient error - even when
+                    // the error would otherwise be treated as transient
+                    Mono<NonTransientPointOperationResult> initialMonoAcrossAllRegions =
+                        callback.apply(clonedOptions, endToEndPolicyConfig, diagnosticsFactory)
+                                .map(response -> new NonTransientPointOperationResult(response));
+
+                        if (logger.isDebugEnabled()) {
+                            monoList.add(initialMonoAcrossAllRegions.doOnSubscribe(c -> logger.debug(
+                                "STARTING to process {} operation in region '{}'",
+                                operationType,
+                                region)));
+                        } else {
+                            monoList.add(initialMonoAcrossAllRegions);
+                        }
+                } else {
+                    clonedOptions.setExcludeRegions(
+                        getEffectiveExcludedRegionsForHedging(
+                            initialRequestOptions.getExcludeRegions(),
+                            orderedApplicableRegionsForSpeculation,
+                            region)
+                    );
+
+                    // Non-Transient errors are mapped to a value - this ensures the firstWithValue
+                    // operator below will complete the composite Mono for both successful values
+                    // and non-transient errors
+                    Mono<NonTransientPointOperationResult> regionalCrossRegionRetryMono =
+                        callback.apply(clonedOptions, endToEndPolicyConfig, diagnosticsFactory)
+                                .map(response -> new NonTransientPointOperationResult(response))
+                                .onErrorResume(
```

Comment:
We have the transformation step from the second `Mono` onwards for non-transient exceptions but not for the first - what is the reasoning behind this?

Reply:
The first one executes exactly as like without availability strategy - if that request fails it means the client retry policy would have done any cross-regional retries - so any outcome of the operation against the initial region (with all preferred regions) is always a non-transient outcome because otherwise our existing retry policies would retry them